js/src/evaluation/_runner.ts

```diff
@@ -0,0 +1,932 @@
+  processedCount = 0;
+  summaryResults: EvaluationResults;
+
+  constructor(experimentManager: _ExperimentManager) {
+    this.manager = experimentManager;
+  }
+
+  get experimentName(): string {
+    return this.manager.experimentName;
+  }
+
+  [Symbol.asyncIterator](): AsyncIterableIterator<ExperimentResultRow> {
+    return this;
+  }
+
+  async next(): Promise<IteratorResult<ExperimentResultRow>> {
+    if (this.processedCount < this.results.length) {
+      const result = this.results[this.processedCount];
+      this.processedCount++;
+      return Promise.resolve({ value: result, done: false });
+    } else {
+      return Promise.resolve({ value: undefined, done: true });
+    }
+  }
+
+  async processData(manager: _ExperimentManager): Promise<void> {
+    for await (const item of manager.getResults()) {
+      this.results.push(item);
+      this.processedCount++;
+    }
+    this.summaryResults = await manager.getSummaryScores();
+  }
+
+  get length(): number {
+    return this.results.length;
+  }
+}
+
+async function _evaluate(
+  target: TargetT | AsyncGenerator<Run>,
+  fields: EvaluateOptions & {
+    experiment?: TracerSession;
+  }
+): Promise<ExperimentResults> {
+  const client = fields.client ?? new Client();
+  const runs = _isCallable(target) ? null : (target as AsyncGenerator<Run>);
+  const [experiment_, newRuns] = await _resolveExperiment(
+    fields.experiment ?? null,
+    runs,
+    client
+  );
+
+  let manager = await new _ExperimentManager({
+    data: fields.data,
+    client,
+    metadata: fields.metadata,
+    experiment: experiment_ ?? fields.experimentPrefix,
+    runs: newRuns ?? undefined,
+  }).start();
+
+  if (_isCallable(target)) {
+    manager = await manager.withPredictions(target as TargetT, {
+      maxConcurrency: fields.maxConcurrency,
+    });
+  }
+  if (fields.evaluators) {
+    manager = await manager.withEvaluators(fields.evaluators, {
+      maxConcurrency: fields.maxConcurrency,
+    });
+  }
+  if (fields.summaryEvaluators) {
+    manager = await manager.withSummaryEvaluators(fields.summaryEvaluators);
+  }
+  // Start consuming the results.
+  const results = new ExperimentResults(manager);
+  await results.processData(manager);
+  return results;
+}
+
+async function _forward(
+  fn: (...args: any[]) => Promise<any> | any,
+  example: Example,
+  experimentName: string,
+  metadata: KVMap,
+  client: Client
+): Promise<_ForwardResults> {
+  let run: BaseRun | null = null;
+
+  const _getRun = (r: RunTree): void => {
+    run = r;
+  };
+
+  const options = {
+    reference_example_id: example.id,
+    on_end: _getRun,
+    project_name: experimentName,
+    metadata: {
+      ...metadata,
+      example_version: example.modified_at
+        ? new Date(example.modified_at).toISOString()
+        : new Date(example.created_at).toISOString(),
+    },
+    client,
+  };
+
+  const wrappedFn = wrapFunctionAndEnsureTraceable(fn, options) as ReturnType<
```

Comment:
This will fail if the person passes in a traceable function right... since the `reference_example_id` won't be bound to the function? We need to guarantee that the run will be correctly traced, so I'd probably always wrap the function since we currently have no way of providing these values at invocation time

Reply:
Fixed to always wrap